src/reactionsystem.jl

```diff
@@ -250,10 +250,17 @@ function ReactionSystem(eqs, iv, states, ps;
     var_to_name = Dict()
     MT.process_variables!(var_to_name, defaults, states′)
     MT.process_variables!(var_to_name, defaults, ps′)
-    MT.collect_var_to_name!(var_to_name, (eq.lhs for eq in observed))
+    MT.collect_var_to_name!(var_to_name, eq.lhs for eq in observed)
 
-    ReactionSystem(eqs′, iv′, states′, ps′, var_to_name, observed, name, systems,
-                   defaults, connection_type, constraints;
+    if constraints !== nothing
+        MT.process_variables!(var_to_name, defaults, get_ps(constraints))

```

Comment:
I think this is a bit trickier / more work than I initially realized -- my apologies for not realizing this sooner!

If we want to maintain consistency with how states, parameters and equations are handled, then `rn.defaults` should not store default values for states/parameters that live in the constraint system (these should be stored in the constraint system). So the `ReactionSystem.defaults` field should only give defaults that actually live within `rn`. To get all the `defaults` we should overload `ModelingToolkit.defaults(...)` to merge in the constraint system's defaults, similar to how we overload `states`, `parameters` and `equations` at:
https://github.com/SciML/Catalyst.jl/blob/6f391eb194ae4aefde1ce851c94391c274c3b021/src/reactionsystem.jl#L305

It probably makes sense to treat `observed` similarly. 

I'm now thinking it might be better to just overload `getproperty` for `ReactionSystem`s to look first in the `ReactionSystem`, and then in the constraint system for symbols. That avoids having a weird hybrid where `var_to_name` merges the `ReactionSystem` and the constraint system together, but all other fields are kept separate between the two systems.

The alternative is to truly treat the constraint subsystem as part of the `ReactionSystem`, but in that case it seems like there is no purpose to having it as its own system (i.e. we should just store all its info in the `ReactionSystem`s fields). I had avoided this approach with the idea that the `ReactionSystem` fields correspond to species and reactions, with other variables and equations stored in the constraint system. It seemed like this would make subsequent analysis tooling easier to handle.

If you want to tackle this please feel free, but otherwise if you let me know I can try to handle this over the next several days. And sorry again for going in a circle on this!

Reply:
Actually, I think your original suggestion of overloading `getvar` is more appropriate to avoid an `ArgumentError` being thrown otherwise.

Reply:
@isaacsas I would be happy to do some more work on this. As you say, it is probably best to simply define some MT methods for ReactionSystem s so that the conceptual separation between the reaction network and the constraints system is maintained.

I will work on those methods presently.

> And sorry again for going in a circle on this!

No worries!

> The alternative is to truly treat the constraint subsystem as part of the ReactionSystem

That is interesting. Might it not be the case that a lot of code could be simplified or removed if that was how it was done? In that case, a `ReactionSystem` might simply be an `AbstractSystem` which can have both `Equation`s and `Reaction`s in the `eqs` field. The trade off would be extra logic in the `convert` functions, I would guess, but on balance it might work out to a leaner codebase.

Reply:
Sounds good.

Merging would reduce having to overload ModelingToolkit functions, but not eliminate it (for example, some of the ModelingToolkit functions assume concrete types for the equation array but we'd have to use `Any`). It also means we no longer know more concretely what are species vs. algebraic variables, or know without searching reactions vs. equations (though this could be handled with a special array type at system construction).

Reply:
Could we not also add some metadata to variables that are introduced in `Reaction`s, to mark them as species? I will at some point explore this option, but for now I am just working on definitions for `MT.defaults(::ReactionSystem)` and `MT.observed(::ReactionSystem)` as a quick solution.

Reply:
Sure, but wouldn't we then have to ask users to manually add that metadata if they are defining `ReactionSystem`s directly and not by the DSL?

In any case, for now I agree that it would be best to stick with the current approach and just add the two needed functions.

Reply:
> In any case, for now I agree that it would be best to stick with the current approach and just add the two needed functions.

Yes this is easier. Should be working now.

> but wouldn't we then have to ask users to manually add that metadata if they are defining ReactionSystems directly and not by the DSL?

I'm not sure about this, but my initial thought was to handle that transparently in the constructor for `Reaction`s. Unless there is some corner cases that go around the `Reaction` constructor but yet still introduce new species, that might work.

And/or as you suggest, a `@species` macro.